### ISSUE TITLE
CSV download for edits

### ISSUE THREAD

**Issue by awolfe76** [cfpb/hmda-platform#726]
We need to provide a way to allow filers to download their edits in CSV format.

We should allow this by:
- [ ] edit type, a single edit type
- [ ] edit type, all edits combined
- [ ] row (depends on #718)

One way of doing this may be to add it to the `/institutions/<institution>/filings/<period>/submissions/<submissionId>/edits` and `/institutions/<institution>/filings/<period>/submissions/<submissionId>/edits/<syntactical|validity|quality|macro>` endpoints by adding a `?format=csv` parameter.

**Comment by nickgrippin** [comment]
How are we flattening the data?  Especially for the "all edit types combined", how are we to differentiate between the edit types?

The way I see it, there are a couple options.  We can start out each row with the edit type:
```
syntactical|S270|false|loanid1|loanid2|loanid3
syntactical|S100|true
validity|210|false|loanid4|loanid5|loanid6|loanid7
macro|Q008
```
Or we could have a row separate each category:
```
syntactical
S270|false|loanid1|loanid2|loanid3
S100|true
validity
210|false|loanid4|loanid5|loanid6|loanid7
macro
Q008
```
Also, are we using the pipe-delimited format? (I assumed so)

**Comment by nickgrippin** [comment]
Another question: are we including persisted justifications from the macro edits?  They're typically returned from the `edits` endpoint, but I'm not sure if I'm comfortable returning them in CSV format.

**Comment by awolfe76** [comment]
> Another question: are we including persisted justifications from the macro edits? They're typically returned from the edits endpoint, but I'm not sure if I'm comfortable returning them in CSV format.

I think we're ok to __not__ include them for now.

> Also, are we using the pipe-delimited format? (I assumed so)

I think comma separated will work better for filers.

---

Your first option works. Another option would be each edit on its own row. Like this (macro would not have a `loanId`):

```
editType, editId, loanId
syntactical, S100, 1234567890
syntactical, S100, 0987654321
macro, q800, 
```

Either way we may run into some [limitations of excel](https://www.google.com/webhp?sourceid=chrome-instant&ion=1&espv=2&ie=UTF-8#q=maximum+columns+in+excel).

And what do the `true` and `false` mean in your first example?

**Comment by nickgrippin** [comment]
I can switch over to comma delimiters.

As for the format, I'm not sure if this will work no matter how we arrange it (unless we include some warning).  Each loan ID having it's own row (for each edit) will definitely run over the limit, and each edit having it's own row has a very high possibility of running over the column limit.

Would it be safe to assume that banks with high numbers of loans won't be using excel?

**Comment by lesgou** [comment]
Per discussion with @nickgrippin:
1. Filers with large number of lar rows will likely already have cleaned up syntactical and validity edits through vendor software. However, there is still likely to be a large number of filers with syntactical and validity edits, and/or quality edits.

Possibly consider:
Adding a warning for filers once the "export/download" button is selected, with a message that the csv will not be able to be opened using excel (for edits that exceed excel's limitations).

2. The csv should be fixed width (1 loan id per row)

**Comment by awolfe76** [comment]
@nickgrippin - can this one be closed?

**Comment by jmarin** [comment]
@nickgrippin can this ticket be closed?

**Comment by awolfe76** [comment]
Closing. CSV downloads are working. If anything comes up we can create a new issue.
